---
title: "LumA vs LumB - DEA"
author: "Poon Chi Lam"
date: "6/18/2019"
output: html_document
---

- Datasets: TCGA BRCA RNA-seq and MetaBric microarray
- Filtering: 50% low-mean genes out
- Feature type: expression values
- Feature selection: differential expression analysis using limma
- Normalization: quantile normalization

One as training data and the other as testing, that is doing twice.

```{R set up}
suppressPackageStartupMessages({
  library(RColorBrewer)
  library(tidyr)
  library(dplyr)
  library(reshape2)
  library(limma)
  library(DESeq2)
  library(edgeR)
  library(ggplot2)
  library(pheatmap)
  library(preprocessCore)
  library(caret)
  library(pROC)
  library(stringr)
})
```


## Load data
```{R load}
data.dir <- '~/GoogleDrive/pjs/kTSP/data'
load(file.path(data.dir, "BRCA.expr.Rdata"))
load(file.path(data.dir, "BRCA.clinic.Rdata"))
load(file.path(data.dir, "MB.expr.Rdata"))
load(file.path(data.dir, "MB.clinic.Rdata"))
```


## Cohort demography
Expression datasets had been intersected.
```{R}
dim(brca.expr); dim(mb.expr)

st.count1 <- brca.clinic %>% group_by(final_assign) %>% summarise(counts = n())
st.count1 <- as.data.frame(st.count1 %>% mutate(data = rep("TCGA", nrow(st.count1))))
st.count2 <- mb.clinic %>% group_by(NOT_IN_OSLOVAL_Pam50Subtype) %>% summarise(counts = n())
st.count2 <- as.data.frame(st.count2 %>% mutate(data = rep("MetaBric", nrow(st.count2))))
colnames(st.count2) <- colnames(st.count1)
st.count <- rbind.data.frame(st.count1, st.count2)

ggplot(st.count, aes(x = final_assign, y = counts, fill = data)) +
  geom_bar(stat = "identity", position=position_dodge()) +
  geom_text(aes(final_assign, counts, label = counts), position=position_dodge(width = 1), vjust = -0.2) +
  scale_y_continuous(limits=c(0, 800)) + 
  theme_bw() +
  labs(title="Sample numbers in each subtype", x="", y="", fill="Data") + 
  scale_fill_brewer(palette="Set1")
```


## Extract LumA & LumB samples
```{R extract}
brca.clin <- brca.clinic[brca.clinic$final_assign %in% c('LumA', 'LumB'),]
brca.expr <- brca.expr[ ,which(colnames(brca.expr) %in% rownames(brca.clin))]
mb.clin <- mb.clinic[mb.clinic$NOT_IN_OSLOVAL_Pam50Subtype %in% c('LumA', 'LumB'),]
mb.expr <- mb.expr[,which(colnames(mb.expr) %in% rownames(mb.clin))]
dim(brca.clin);dim(brca.expr);dim(mb.clin);dim(mb.expr)
```


## Overview of expression
```{R overview, fig.width=12}
# Create a color vector containing up to 74 colour hexcodes
getColors <- function(n) {
  col <- brewer.pal.info[brewer.pal.info$category=='qual', ] # get max. 74 colours
  col_vector <- unlist(mapply(brewer.pal, col$maxcolors, rownames(col)))
  ifelse (n > length(col_vector), 
          vec <- sample(col_vector, n, replace=T),
          vec <- sample(col_vector, n, replace=F)
  )
  vec
}


# Density plot
dens.plot <- function(table, colVec, yrange) {
  d <- plot(density(table[, 1]), col=colVec[1], 
          lwd=2, las=2, ylim=yrange, main="", xlab="") +
      abline(v=0, lty=3) + title(xlab="expr values") +
      for (i in 2:ncol(table)) {
        den <- density(table[, i])
        lines(den$x, den$y, col=colVec[i], lwd=2)
      } 
  d
}

par(mfrow=c(1, 2), las=1)
dens.plot(brca.expr, getColors(ncol(mb.expr)), c(0, 0.6))
dens.plot(mb.expr, getColors(ncol(brca.expr)), c(0, 0.6))
```


## TCGA as testing
### Filter low mean genes
```{R lowmean}
lowMeans.filter <- function(cal.mat, percent) {
  means <- data.frame(means = rowMeans(cal.mat), ranks = rank(rowMeans(cal.mat)))
  cuts <- quantile(means$ranks, probs = seq(0, 1, 0.05))
  
  # Top n% genes (rank() is from the smallest to largest)
  cut.percent <- paste0((100 - percent), "%")
  means <- means[which(means$ranks >= cuts[cut.percent]), ]
  sub.genes <- rownames(means)
  sub.genes
}

tcga.sub <- lowMeans.filter(brca.expr, 50)
length(tcga.sub)
data.train <- brca.expr[which(rownames(brca.expr) %in% tcga.sub),]
data.test <- mb.expr[which(rownames(mb.expr) %in% tcga.sub),]
lb.train <- factor(brca.clin$final_assign)
lb.test <- factor(mb.clin$NOT_IN_OSLOVAL_Pam50Subtype)
```


### DE analysis - TCGA
Use **limma-trend** here, the mean-variance relationship is modelled with an empirical Bayes prior trend, concording with those used in microarrary, since the library sizes in TCGA samples are usually consistent. 
```{R dea}
type <- lb.train
design <- model.matrix(~ type)
head(design, 5)
type1 <- 'typeLumB'

# fitting
vfit <- lmFit(data.train, design)
efit <- eBayes(vfit, trend = T)
plotSA(efit, main="Final model: Mean-variance trend")

# Top table
tt <- topTable(efit, number = Inf, coef = type1)
head(tt, 10)

# Volcano & MA plot
limma::volcanoplot(efit, coef = type1, highlight = 5, names = rownames(efit))
limma::plotMA(efit, main = "TCGA LumA vs LumB")

# Significant gene numbers
nrow(tt[(tt$adj.P.Val < 0.01),])
nrow(tt[(tt$adj.P.Val < 0.01) & (abs(tt$logFC) > log2(1.5)), ])
DEG <- rownames(tt[(tt$adj.P.Val < 0.01), ])
# sDEG <- rownames(tt[(tt$adj.P.Val < 0.01) & (abs(tt$logFC) > log2(1.5)), ])
# nDEG <- rownames(tt)[!rownames(tt) %in% DEG]
# nsDEG <- rownames(tt)[!rownames(tt) %in% sDEG]

data.train <- data.train[which(rownames(data.train) %in% DEG),]
data.test <- data.test[which(rownames(data.test) %in% DEG),]
```

### Heatmap
```{R heatmap}
draw_heatmap <- function(matrix, topTable, phenoDF, list) {
  matrix <- matrix[list,]
  hm_cdr <- phenoDF %>% select(final_assign)
  rownames(hm_cdr) <- colnames(matrix)
  colnames(hm_cdr)  <- c('type')
  type <- c("#99a599", "#37637f")
  names(type) <- unique(hm_cdr$type)
  anno_colors <- list(type = type)
  h <- pheatmap(matrix, annotation_col=hm_cdr, annotation_colors=anno_colors, 
                labels_row = list, show_colnames = F)
  h
}

draw_heatmap(data.train, tt, brca.clin, rownames(tt)[1:30])
```

Those selected DEGs are used as features of algorithms in the next step.


### Quantile normalization
The one as training is the target, then the training one is to be normalized.
```{R QN}
countDis <- function(countMat1, countMat2, labels) {
  ga.count1 <- countMat1 %>% gather(ID)
  ga.count1$Data <- labels[1]
  ga.count2 <- countMat2 %>% gather(ID)
  ga.count2$Data <- labels[2]
  
  count.df <- rbind.data.frame(ga.count1, ga.count2)
  mu <- count.df %>%
    select(value, Data) %>%
    group_by(Data) %>%
    summarise(mean.count = mean(value, na.rm = T))
  
  ggplot(count.df, aes(x=value, color=Data, fill=Data)) + 
    geom_histogram(aes(y=..density..), alpha=.5, binwidth=0.6) +
    geom_density(alpha=0.2) +
    geom_vline(data=mu, aes(xintercept=mean.count, color=Data), linetype='dashed') + 
    scale_fill_brewer(palette="Set1") + 
    scale_color_brewer(palette="Set1") + theme_bw() +
    labs(x='Count', y='Density')
}

# Counts before quantile normalization
countDis(data.train, data.test, c('TCGA', 'MetaBric'))

# normalize
nm.data.train <- normalize.quantiles(as.matrix(data.train), copy=F)
target <- normalize.quantiles.determine.target(nm.data.train)
nm.data.test <- normalize.quantiles.use.target(as.matrix(data.test), target)

nm.data.train <- as.data.frame(nm.data.train)
nm.data.test <- as.data.frame(nm.data.test)
rownames(nm.data.test) <- rownames(data.test)
colnames(nm.data.test) <- colnames(data.test)

# After normalization
countDis(nm.data.train, nm.data.test, c('TCGA', 'MetaBric'))

save(tt, nm.data.train, nm.data.test, lb.train, lb.test, file=file.path(data.dir, "tcga.DEG.Rdata"))
```


## MetaBric as training
### Filter low mean genes
```{R}
mb.sub <- lowMeans.filter(mb.expr, 50)
length(mb.sub)
data.train <- mb.expr[which(rownames(mb.expr) %in% mb.sub),]
data.test <- brca.expr[which(rownames(brca.expr) %in% mb.sub),]
lb.train <- factor(mb.clin$NOT_IN_OSLOVAL_Pam50Subtype)
lb.test <- factor(brca.clin$final_assign)
```


### DE analysis - MetaBric
Using limma-trend here as well.
```{R}
type <- lb.train
design <- model.matrix(~ type)
head(design, 5)
type1 <- 'typeLumB'

vfit <- lmFit(data.train, design) 
efit <- eBayes(vfit, trend = T)
plotSA(efit, main="Final model: Mean-variance trend")

# Top table
tt <- topTable(efit, number = Inf, coef = type1)
head(tt, 10)

# Volcano & MA plot
limma::volcanoplot(efit, coef = type1, highlight = 5, names = rownames(efit))
limma::plotMA(efit, main = "MetaBric LumA vs LumB")

# Significant gene numbers
nrow(tt[(tt$adj.P.Val < 0.01), ])
nrow(tt[(tt$adj.P.Val < 0.01) & (abs(tt$logFC) > log2(1.5)), ])
DEG <- rownames(tt[(tt$adj.P.Val < 0.01), ])
# nDEG <- rownames(tt)[!rownames(tt) %in% DEG]

data.train <- data.train[which(rownames(data.train) %in% DEG),]
data.test <- data.test[which(rownames(data.test) %in% DEG),]
```

### Heatmap
```{R}
draw_heatmap <- function(matrix, topTable, phenoDF, list) {
  matrix <- matrix[list,]
  hm_cdr <- phenoDF %>% select(NOT_IN_OSLOVAL_Pam50Subtype)
  rownames(hm_cdr) <- colnames(matrix)
  colnames(hm_cdr)  <- c('type')
  type <- c("#99a599", "#37637f")
  names(type) <- unique(hm_cdr$type)
  anno_colors <- list(type = type)
  h <- pheatmap(matrix, annotation_col=hm_cdr, annotation_colors=anno_colors, 
                labels_row = list, show_colnames = F)
  h
}

draw_heatmap(data.train, tt, mb.clin, rownames(tt)[1:30])
```


### Quantile normalization
The one as training is the target, then the training one is to be normalized.
```{R}
# Counts before quantile normalization
countDis(data.test, data.train, c('TCGA', 'MetaBric'))

# normalize
nm.data.train <- normalize.quantiles(as.matrix(data.train), copy=F)
target <- normalize.quantiles.determine.target(nm.data.train)
nm.data.test <- normalize.quantiles.use.target(as.matrix(data.test), target)

nm.data.train <- as.data.frame(nm.data.train)
nm.data.test <- as.data.frame(nm.data.test)
rownames(nm.data.test) <- rownames(data.test)
colnames(nm.data.test) <- colnames(data.test)

# After normalization
countDis(nm.data.test, nm.data.train, c('TCGA', 'MetaBric'))

save(tt, nm.data.train, nm.data.test, lb.train, lb.test, file=file.path(data.dir, "mb.DEG.Rdata"))
```

