---
title: "Breast Cancer multiclass - baseline testing"
author: "Poon Chi Lam"
date: "8/5/2019"
output: html_document
---

- Datasets: TCGA BRCA RNA-seq and MetaBric microarray
- Filtering: 30% low-mean genes out
- Feature type: expression values
- Feature selection: differential expression analysis using limma-trend
- Normalization: QN, FSQN, TDM
- Algorithm: random forest

- Seletcted Feature sets:

TCGA: top 4% genes
MetaBric: top 4% genes


```{R setup}
suppressPackageStartupMessages({
  library(tidyr)
  library(dplyr)
  library(reshape2)
  library(ggplot2)
  library(ggpubr)
  library(caret)
  library(pROC)
  library(stringr)
  library(randomForest)
})
source('~/ktsp/script/function/plot.utils.R')
source('~/ktsp/script/function/cross.platform.norm.R')
```


### load data
```{R}
load('~/ktsp/data/multiclass/realdata/imbalanced/BRCA.multi.Rdata')
load('~/ktsp/data/multiclass/realdata/imbalanced/MB.multi.Rdata')

lb.brca <- factor(mul.brca.clin$final_assign)
lb.mb <- factor(mul.mb.clin$Pam50Subtype)

# genes with p-val
load('~/ktsp/data/multiclass/realdata/imbalanced/tcga.fscore.Rdata')
brca.train <- mul.brca.expr[match(features[["sub.04"]], rownames(mul.brca.expr)), ]
mb.test <- mul.mb.expr[match(features[["sub.04"]], rownames(mul.mb.expr)), ]

load('~/ktsp/data/multiclass/realdata/imbalanced/mb.fscore.Rdata')
mb.train <- mul.mb.expr[match(features[["sub.04"]], rownames(mul.mb.expr)), ]
brca.test <- mul.brca.expr[match(features[["sub.04"]], rownames(mul.brca.expr)), ]
```


### Organize functions
```{R}
source('~/ktsp/script/function/multi.rf.test.R')
```


## TCGA as training
### Cross-platform normalization
```{R}
qn.mb.test <- cross.platform.norm(data.train=brca.train, 
                                  data.test=mb.test,
                                  method='QN')

tdm.mb.test <- cross.platform.norm(data.train=brca.train, 
                                  data.test=mb.test,
                                  method='TDM')
```


### Train model and testing
```{R}
tcga.model <- multi.model(as.data.frame(t(brca.train)), lb.brca)

brca.res <- list()
for (i in c('QN', 'TDM')) {
  if (i == 'QN') {
    nm.test <- qn.mb.test
  } else if (i == 'TDM') {
    nm.test <- tdm.mb.test
  }
  
  brca.res[[i]] <- multi.rf.test(testdata=as.data.frame(t(nm.test)), 
                             testlabel=lb.mb, 
                             model=tcga.model)
  print(brca.res[[i]][["cm"]])
}

save(brca.res, file='~/ktsp/data/multiclass/realdata/imbalanced/BRCA.RF.test.Rdata')
```



## METABRIC as training
### Cross-platform normalization
```{R}
qn.brca.test <- cross.platform.norm(data.train=mb.train, 
                                    data.test=brca.test,
                                    method='QN')

tdm.brca.test <- cross.platform.norm(data.train=mb.train, 
                                     data.test=brca.test,
                                     method='TDM')

```


### Train model and testing
```{R}
mb.model <- multi.model(as.data.frame(t(mb.train)), lb.mb)

mb.res <- list()
for (i in c('QN', 'TDM')) {
  if (i == 'QN') {
    nm.test <- qn.brca.test
  } else if (i == 'TDM') {
    nm.test <- tdm.brca.test
  }
  
  mb.res[[i]] <- multi.rf.test(testdata=as.data.frame(t(nm.test)), 
                             testlabel=lb.brca, 
                             model=mb.model)
  print(mb.res[[i]][["cm"]])
}

save(mb.res, file='~/ktsp/data/multiclass/realdata/imbalanced/MB.RF.test.Rdata')
```



### Visualization
```{R, fig.width=12, fig.height=4}
## TCGA TESTING
norm.method <- c('QN', 'TDM')
final.plot <- data.frame()
for (norm in norm.method) {
  df <- data.frame(ACC=0, ACCub=0, ACClb=0, YD=0)
  list <- brca.res[[norm]]
  df$ACC <- list[['ACC']]
  df$ACCub <- list[["cm"]][["overall"]][["AccuracyUpper"]]
  df$ACClb <- list[["cm"]][["overall"]][["AccuracyLower"]]
  df$YD <- list[["avg.yd"]]
  df$Method <- rep(norm, nrow(df))
  final.plot <- rbind(final.plot, df)
}
final.plot$Method <- paste0('RF+', final.plot$Method)

# load ROML result
load('~/ktsp/data/multiclass/realdata/imbalanced/BRCA.ROML.test.m1.Rdata')
final.plot[3,]$ACC <- brca.test.res[["20"]][["ACC"]]
final.plot[3,]$ACCub <- brca.test.res[["20"]][["cm"]][["overall"]][["AccuracyUpper"]]
final.plot[3,]$ACClb <- brca.test.res[["20"]][["cm"]][["overall"]][["AccuracyLower"]]
final.plot[3,]$YD <- brca.test.res[["20"]][["avg.yd"]]
final.plot[3,]$Method <- 'ROML-M1'

load('~/ktsp/data/multiclass/realdata/imbalanced/BRCA.ROML.test.m2.Rdata')
final.plot[4,]$ACC <- brca.test.res[["300"]][["ACC"]]
final.plot[4,]$ACCub <- brca.test.res[["300"]][["cm"]][["overall"]][["AccuracyUpper"]]
final.plot[4,]$ACClb <- brca.test.res[["300"]][["cm"]][["overall"]][["AccuracyLower"]]
final.plot[4,]$YD <- brca.test.res[["300"]][["avg.yd"]]
final.plot[4,]$Method <- 'ROML-M2'

final.plot$YD <- as.numeric(final.plot$YD)
final.plot$Method <- factor(final.plot$Method, levels=as.character(final.plot$Method))
print(final.plot)

# TEST acc
g1 <- ggplot(final.plot, aes(x=Method, y=ACC, group=Method, color=Method)) + 
  geom_point(size=4) + labs(y='Accuracy', x='') + ggtitle(paste0('TCGA TEST ACC')) +
  geom_errorbar(aes(ymin=ACClb, ymax=ACCub), width=.05,
             position=position_dodge(0.05)) +
  #scale_x_discrete(limits = norm.method) +
  scale_y_continuous(breaks = seq(0,1, by=0.1), limits = c(0,1)) +
  scale_color_brewer(palette='Set2') +
  theme_bw()


# TEST youden
g2 <- ggplot(final.plot, aes(x=Method, y=YD, group=Method, color=Method)) + 
  geom_point(size=4) + labs(y='Youden Index', x='') + 
  ggtitle('TCGA Youden Index') +
  #geom_errorbar(aes(ymin=meanYD-sdYD, ymax=meanYD+sdYD), width=.2,
            # position=position_dodge(0.05)) +
  #scale_x_discrete(limits = norm.method) +
  scale_y_continuous(breaks = seq(0,1, by=0.1), limits = c(0,1)) +
  scale_color_brewer(palette='Set2') +
  theme_bw()

ggarrange(g1, g2, ncol=2, nrow=1, common.legend = T, legend = 'right')
```


```{R, fig.width=12, fig.height=4}
## MB TESTING
final.plot <- data.frame()
for (norm in norm.method) {
  df <- data.frame(ACC=0, ACCub=0, ACClb=0, YD=0)
  list <- mb.res[[norm]]
  df$ACC <- list[['ACC']]
  df$ACCub <- list[["cm"]][["overall"]][["AccuracyUpper"]]
  df$ACClb <- list[["cm"]][["overall"]][["AccuracyLower"]]
  df$YD <- list[["avg.yd"]]
  df$Method <- rep(norm, nrow(df))
  final.plot <- rbind(final.plot, df)
}
final.plot$Method <- paste0('RF+', final.plot$Method)


# load ROML result
load('~/ktsp/data/multiclass/realdata/imbalanced/MB.ROML.test.m1.Rdata')
final.plot[3,]$ACC <- mb.test.res[["20"]][["ACC"]]
final.plot[3,]$ACCub <- mb.test.res[["20"]][["cm"]][["overall"]][["AccuracyUpper"]]
final.plot[3,]$ACClb <- mb.test.res[["20"]][["cm"]][["overall"]][["AccuracyLower"]]
final.plot[3,]$YD <- mb.test.res[["20"]][["avg.yd"]]
final.plot[3,]$Method <- 'ROML-M1'

load('~/ktsp/data/multiclass/realdata/imbalanced/MB.ROML.test.m2.Rdata')
final.plot[4,]$ACC <- mb.test.res[["300"]][["ACC"]]
final.plot[4,]$ACCub <- mb.test.res[["300"]][["cm"]][["overall"]][["AccuracyUpper"]]
final.plot[4,]$ACClb <- mb.test.res[["300"]][["cm"]][["overall"]][["AccuracyLower"]]
final.plot[4,]$YD <- mb.test.res[["300"]][["avg.yd"]]
final.plot[4,]$Method <- 'ROML-M2'

final.plot$YD <- as.numeric(final.plot$YD)
final.plot$Method <- factor(final.plot$Method, levels=as.character(final.plot$Method))
print(final.plot)

# TEST acc
g1 <- ggplot(final.plot, aes(x=Method, y=ACC, group=Method, color=Method)) + 
  geom_point(size=4) + labs(y='Accuracy', x='') + ggtitle(paste0('METABRIC TEST ACC')) +
  geom_errorbar(aes(ymin=ACClb, ymax=ACCub), width=.05,
             position=position_dodge(0.05)) +
  #scale_x_discrete(limits = norm.method) +
  scale_y_continuous(breaks = seq(0,1, by=0.1), limits = c(0,1)) +
  scale_color_brewer(palette='Set2') +
  theme_bw()


# TEST youden
g2 <- ggplot(final.plot, aes(x=Method, y=YD, group=Method, color=Method)) + 
  geom_point(size=4) + labs(y='Youden Index') + 
  ggtitle('METABRIC Youden Index') +
  #geom_errorbar(aes(ymin=meanYD-sdYD, ymax=meanYD+sdYD), width=.2,
            # position=position_dodge(0.05)) +
  #scale_x_discrete(limits = norm.method) +
  scale_y_continuous(breaks = seq(0,1, by=0.1), limits = c(0,1)) +
  scale_color_brewer(palette='Set2') +
  theme_bw()

ggarrange(g1, g2, ncol=2, nrow=1, common.legend = T, legend = 'right')
```






