---
title: "Balanced LumA vs LumB - Random Forest Cross Validation"
author: "Poon Chi Lam"
date: "7/24/2019"
output: html_document
---

- Datasets: TCGA BRCA RNA-seq and MetaBric microarray
- Filtering: 30% low-mean genes out
- Feature type: expression values
- Feature selection: differential expression analysis using limma-trend
- Algorithm: random forest

- Feature sets:

Top 50, 100, 300, 500, 1000, 5000 DEGs (q-val < 0.01)


-> Select the best feature set by the performance of cross-validation.


One as training data and the other as testing, that is doing twice.
```{R setup}
suppressPackageStartupMessages({
  library(tidyr)
  library(dplyr)
  library(reshape2)
  library(ggplot2)
  library(ggpubr)
  library(caret)
  library(pROC)
  library(stringr)
  library(preprocessCore)
  library(randomForest)
})
source('~/ktsp/script/function/plot.utils.R')
source('~/ktsp/script/function/lowMeans.filter.R')
source('~/ktsp/script/function/cross.platform.norm.R')
source('~/ktsp/script/function/binary.cv.R')
```


## Load data
```{R load}
data.dir <- '~/ktsp/data/binary/realdata/balanced'
load(file.path(data.dir, "TCGA.balance.LumAB.Rdata"))
load(file.path(data.dir, "MB.balance.LumAB.Rdata"))
lb.brca <- factor(brca.clin.sub$final_assign)
lb.mb <- factor(mb.clin.sub$Pam50Subtype)
```



## TCGA as training
Top 50, 100, 300, 500, 1000, 5000 DEGs (q-val < 0.01)
```{R tcga, eval=T}
gene.remain <- lowMeans.filter(brca.expr.sub, 70)
brca.train <- brca.expr.sub[match(gene.remain, rownames(brca.expr.sub)), ]

# load("~/ktsp/data/binary/realdata/balanced/baseline/tcga.DEG.Rdata")
# brca.train <- brca.expr.sub[match(rownames(tt.brca), rownames(brca.expr.sub)),]
# brca.train <- normalize.quantiles(as.matrix(brca.train), copy=F)
# colnames(brca.train) <- colnames(brca.expr.sub)
# rownames(brca.train) <- rownames(tt.brca)


# Run cross validation
deg.num <- c(50, 100, 300, 500, 1000, 5000)
tcga.cv <- bi.cv(time=10, 
                 data.train=brca.train, 
                 lb.train=lb.brca, 
                 feature.num=deg.num, 
                 baseline=TRUE)
save(tcga.cv, file=file.path(data.dir, 'baseline/result/new.tcga.cv.rf.Rdata'))
```




## MetaBric as training

```{R, eval=T}
# load("~/ktsp/data/binary/realdata/balanced/baseline/mb.DEG.Rdata")
# mb.train <- mb.expr.sub[match(rownames(tt.mb), rownames(mb.expr.sub)), ]
# mb.train <- normalize.quantiles(as.matrix(mb.train), copy=F)
# colnames(mb.train) <- colnames(mb.expr.sub)
# rownames(mb.train) <- rownames(tt.mb)

gene.remain <- lowMeans.filter(mb.expr.sub, 70)
mb.train <- mb.expr.sub[match(gene.remain, rownames(mb.expr.sub)), ]
```


## Cross-validation
```{R, eval=T}
mb.cv <- bi.cv(time=10, 
               data.train=mb.train, 
               lb.train=lb.mb, 
               feature.num=deg.num, 
               baseline=TRUE)
save(mb.cv, file=file.path(data.dir, 'baseline/result/new.mb.cv.rf.Rdata'))
```


## Visulization
```{R vis, fig.width=12, fig.height=4}
draw.perf <- function(name) {
  list <- get(paste0(name,'.cv'))
  df <- list[['res.df']]
  df$feature.num <- factor(df$feature.num, levels = df$feature.num)
  # CV bac
  g1 <- ggplot(df, aes(x=feature.num, y=MEANbac, group=1)) + 
        geom_line(size=1.1, color='darkgrey') + 
        geom_point(size=1.1) + labs(y='BAC') + ggtitle(paste0(name, ' CV BAC')) +
        geom_errorbar(aes(ymin=MEANbac-SDbac, ymax=MEANbac+SDbac), width=.2,
                   position=position_dodge(0.05), color='darkgrey') +
        scale_x_discrete(limits = df$feature.num) +
        scale_y_continuous(breaks = seq(0.6, 1, by=0.1), limits = c(0.6,1)) +
        theme_bw()
  
  # youden index
  g2 <- ggplot(df, aes(x=feature.num, y=MEANyd, group=1)) + 
        geom_line(size=1.1, color='steelblue') + 
        geom_point(size=1.1) + labs(y='Youden Index') + ggtitle(paste0(name, ' CV Youden Index')) +
        geom_errorbar(aes(ymin=MEANyd-SDyd, ymax=MEANyd+SDyd), width=.2,
                   position=position_dodge(0.05), color='steelblue') +
        scale_x_discrete(limits = df$feature.num) +
        scale_y_continuous(breaks = seq(0.6, 1, by=0.1), limits = c(0.6,1)) +
        theme_bw()
  print(ggarrange(g1, g2, ncol=2, nrow=1, common.legend = T, legend = 'right'))
}


draw.perf('tcga')
draw.perf('mb')
```